contracts/os/andromeda-kernel/src/execute.rs

```diff
@@ -918,251 +918,260 @@ pub fn unset_env(execute_ctx: ExecuteContext, variable: String) -> Result<Respon
-//         Ok::<String, ContractError>(ics20_channel)
-//     } else {
-//         return Err(ContractError::InvalidPacket {
-//             error: Some(format!("Channel not found for chain {chain}")),
-//         });
-//     }?;
-//     deps.api.debug(&format!("info.funds: {:?}", info.funds));
-//     deps.api.debug(&format!("funds: {:?}", funds));
-//     ensure!(
-//         funds.len() == 1,
-//         ContractError::InvalidFunds {
-//             msg: "Number of funds should be exactly one".to_string()
-//         }
-//     );
-//     let coin = funds
-//         .first()
-//         .ok_or(ContractError::InvalidPacket {
-//             error: Some("Transfer funds must contain funds in the AMPMsg".to_string()),
-//         })?
-//         .clone();
-
-//     let msg = IbcMsg::Transfer {
-//         channel_id: channel.clone(),
-//         to_address: channel_info.kernel_address.clone(),
-//         amount: coin.clone(),
-//         timeout: env.block.time.plus_seconds(PACKET_LIFETIME).into(),
-//     };
-//     let mut resp = Response::default();
-
-//     // Save execute msg, to be loaded in the reply
-//     PENDING_MSG_AND_FUNDS.save(
-//         deps.storage,
-//         &Ics20PacketInfo {
-//             sender: info.sender.into_string(),
-//             recipient: recipient.clone(),
-//             message: message.clone(),
-//             funds: coin,
-//             channel: channel.clone(),
-//             pending: false,
-//         },
-//     )?;
-//     resp = resp.add_submessage(SubMsg {
-//         id: ReplyId::IBCTransfer.repr(),
-//         msg: CosmosMsg::Ibc(msg),
-//         gas_limit: None,
-//         reply_on: cosmwasm_std::ReplyOn::Always,
-//     });
-//     Ok(resp
-//         .add_attribute(format!("method:{sequence}"), "execute_transfer_funds")
-//         .add_attribute(format!("channel:{sequence}"), channel)
-//         .add_attribute("receiving_kernel_address:{}", channel_info.kernel_address)
-//         .add_attribute("chain:{}", chain))
-// }
+    // Create and add routing information
+    let hop = create_cross_chain_hop(
+        channel,
+        current_chain,
+        destination_chain,
+        ctx.get_origin(),
+        username.map(AndrAddr::from_string),
+    );
+    ctx.add_hop(hop);
+
+    // Assemble the final packet
+    Ok(AMPPkt::new_with_ctx(ctx, vec![amp_msg]))
+}
+
+fn create_cross_chain_hop(
+    channel: &str,
+    current_chain: String,
+    destination_chain: &String,
+    origin_address: String,
+    username: Option<AndrAddr>,
+) -> CrossChainHop {
+    CrossChainHop {
+        username,
+        address: origin_address,
+        from_chain: current_chain,
+        to_chain: destination_chain.clone(),
+        funds: vec![], // Assuming funds are not handled here as per the given instructions
+        channel: channel.to_string(),
+    }
+}
+
+pub(crate) fn handle_ibc_direct(
+    deps: DepsMut,
+    info: MessageInfo,
+    env: Env,
+    ctx: Option<AMPPkt>,
```

Comment:
I think this should be renamed to pkt or packet, since it's not `Option<AMPCtx>`